### ISSUE TITLE
[no-restricted-globals] False positive on Typescript types

### ISSUE THREAD

**Issue by buksy90** [typescript-eslint/typescript-eslint#4715]
**no-restricted-globals** False positive on Typescript types

I have checked and **there is no** typescript extension rule @typescript-eslint/no-restricted-globals

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```
// Correct error on Promise.resolve
// Incorrect error on Promise<any>
const x: Promise<any> = Promise.resolve();


// Correct no error on window.Promise.resolve
// Incorrect error on Promise<any>
const x: Promise<any> = window.Promise.resolve();
```

[Here is an example in playground](https://typescript-eslint.io/play/#ts=4.6.2&sourceType=module&code=PTAEGEHsCdoUwMYBdR1jUkB2oAK1IBbASwGc4A6eUyAGwDc4BYAKBFAEksEZ5lV00TDnxEycADwBDLAE8AfKx5ZSKAB4AuPARLlpc+aAC82seSpwaDOAAoAlAG5WztmCixEKLJAEEh2UAB3YiwAE0hAilFdSmo6RlZ2Lh4PfjQ-YVMY-QUlbFVQTSzxHMMTYLCIqJ1xCytGeycWIA&rules=N4Igdg9gtATgpgZwC4wJYGMlwCZQOYA2EARgIYEIgBcA2iHDDBDCADShikC2c1ICcJABVUPCAFckIAL7tw3XlX6CAklx7ZUpLDLmcefeAEdxiJAEEwo7aghgAYjAW6OzpUXQ27L+QaUALVGRmAE8ffUUQAAUmLiDeaQBdaSA&tsConfig=N4XyA)

```JSON
{
  "rules": {
    "no-restricted-globals": ['error', { name: 'Promise' }]
  }
}
```

**tsconfig.json**
```
{
    "include": [
      "src/**/*.ts",
    ],
    "exclude": ["node_modules"],
    "compilerOptions": {
      "target": "esnext",
      "module": "esnext",
      "strict": true,
      "declaration": true,
      "declarationDir": "dist",
      "outDir": "dist",
      "importHelpers": true,
      "moduleResolution": "node",
      "emitDecoratorMetadata": true,
      "experimentalDecorators": true,
      "esModuleInterop": false,
      "allowSyntheticDefaultImports": true,
      "noImplicitReturns": true,
      "sourceMap": true,
      "lib": ["esnext", "dom", "dom.iterable", "scripthost"],
      "typeRoots": ["./node_modules/@types", "./typings"],
      "baseUrl": ".",
      "paths": {
        "@/*": ["src/*"],
        "@test/*": ["srctest/*"]
      }
    }
  }

```

**Additional info**
My goal is to prohibit usage of global `Promise` object and instead I want this object to be always referenced from `window.Promise`. This works well with javascript, but the problem comes when `Promise` is used as type in typescript files.

**Expected Result**
Get error only on usage of global Promise as value, not type.

**Actual Result**
Getting error on global usage of both Promise as value and type.

**Versions** (copied versions from online typescript eslint playground)

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.16.0` |
| `TypeScript`                       | `4.6.2` |
| `ESLint`                           | `8.1.0` |

**Comment by bradzacher** [comment]
This is really "working as expected" because it's all just globally named variables. Esp. In this case where `Promise` is technically a global class name - I.e. the name is both a type and a value.

I can see why one might like to separate the two. It's a ***very*** rare usecase to want this though.

Why are you looking to ban the global promise type? By definition it should always be the same as `window.Promise`.

**Comment by buksy90** [comment]
Okay thank you :) ..

My intention here was to make sure functions from proper environment are referenced in project. 
Take a `setTimeout` as an example, if you reference it in your .ts files, there is NodeJs environment overload used by default. It is returning `NodeJS.Timeout` type instead of number. I know how to reference version for browser environment, but I noticed many devs on my project are using this NodeJS type when declaring variables.
So I decided to ban usage of `setTimeout` function as global and instead devs have to reference it from `window` object.
Another advantage of this is that it is clearer for devs how to stub this function in tests when they see it is defined on `window` object.

So similarly, I wanted to ban couple of more global functions to make sure they are always referenced from browser environment specific `window` object instead of nodeJs environment specific `global`.

**Comment by bradzacher** [comment]
IMO forcing your users to specifically access everything via `global` or `window` is going to be cumbersome and somewhat painful for users. It will create an inconsistency in your codebase because *some* of your globals will be directly referenced and others will be accessed via `window`.
Options for fixing this at the root cause rather than working around it (in no particular order):
- You could solve this problem by using [`ban-types`](https://typescript-eslint.io/rules/ban-types) to ban `NodeJS.Timeout` in your non-node files.
   - [This is the simplest solution which requires no changes at all.](https://typescript-eslint.io/play/#ts=4.6.2&sourceType=module&code=PQKgBApgzgNglgOwC5gAJIJ4AdoGMBOcWSAtNPMsAEYCGCJmOUAXGANoDkE++A9vhwA0YAN5hG0VmI4A5XgBMIAKQDKAOgAqcALYReAVyQdWs3ikRgA7hCpgAZnBjQwURzAwBCDmAC+PgLpgIMAAUCHAwOK8YNo0ANYQ4gAWiVgwNBgA5nz6CPJgMLw0+UgpYAgKEABWUOLY0CE6WPwoHHZQHADcYbi8CFAoAB6scoqqmjp6hmAAvC4QSFq6BkgAFKsAlLMAfKI+wgAMG51AA&rules=N4XyA&tsConfig=N4XyA)
   - This is the best option as it means you can target the fix.
- If you're working in a monorepo - consider installing `@types/node` just in the package(s) that need it, rather than in the project root.
   - This will prevent your web projects from accessing the types at all!
- If you use the [`types`](https://www.typescriptlang.org/tsconfig#types) tsconfig option you can narrow down and disallow the nodejs types from being used for specific projects.
   - This is pretty cumbersome to do as it's an allowlist, not a blocklist - but it could work if your list of deps is static.

I don't think it's worth the effort to build an extension rule to support this very rare usecase, given there are the above workarounds. For that reason I'm going to close this.
